### PR TITLE
Full Stack: Gareth Welch

### PR DESCRIPTION
**Pull request by GarethPW**
## 0. Housekeeping

I decided to stick with Ruby 3.0 for the time being to minimise gem incompatibilities where possible. It wouldn't surprise me if Rails 6.1 can run just fine on newer Ruby releases, but the possibility of having to spend lots of time on that didn't seem worth it. Legacy app upgrades can be a project of their own, and whilst upgrading from 3.0 should be a priority I consider it out of scope here.

Other dependencies which required attention:

* mimemagic (-> 0.3.10) – yanked due to licensing fiasco
* Webpacker (-> 5.2.2) – indirect Python 2 dependency!
* bootsnap (-> 1.18.3) – Ruby 3.0.3+ incompatibility
* REXML – not a default gem in Ruby 3.0

Additional details can be found in respective commit messages.

## 1. Review Creation

Added a form, styling, specs, and Sidekiq worker so reviews can be left asychronously. I've opted to postpone adding star icons for the time being in order to deliver an MVP. However, I think an icon font like bootstrap-icons would be a good choice for this down the line.

In refactoring this code, I removed an incomplete feature which theoretically allowed reviewers to submit their own tags. In practice, this feature never would have worked so it didn't make sense to keep it during the refactoring process.

During this challenge, I added Factory Bot to the project bundle to make generating records easier in specs.

45f973b6bf744a48bc665c7d6ff8814a

## 2. Review Index

Revamped the index page with styling, server-side pagination, and a barebones system spec.

It wasn't feasible to implement true AJAX pagination before the deadline, so I compromised with a server-side solution using the will_paginate gem (chosen out of familiarity). Whilst not ideal, this works as a trade-off for product pagination because Turbolinks speeds up navigation with implicit AJAX/XHR loading anyway. 

I expect a different approach will be necessary for pagination of individual review lists. In the meantime, I have increased the number of listed reviews per product to compensate.

0a762cb1421b1d2cf036539e89cb680d

## 3. Review Statistics

I've added an initial, experimental version of the monthly stats page at `/reviews/stats`. No specs for this exist yet so it isn't guaranteed stable. No styling has been added yet either.

---

This is as much as I was able to complete within the time provided. In each challenge, I tried to strike a balance between functionality, styling, and testing since each is important for different reasons. I could have divided time between the three challenges better.

### CODE REVIEW

File: tasks/fullStack/app/helpers/reviews_helper.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      .limit(10)
+  end
+
+  # Paginated products whose reviews will be displayed
+  def reviewed_products
+    @products ||= shop
+      .products
+      .includes(:reviews)
+      .order(created_at: :desc)
+      .paginate(page: params[:page], per_page: 2)
```
Comment (dmut) [inline tasks/fullStack/app/helpers/reviews_helper.rb:28]: would be good if `per_page` could also be sent as `params`

**Comment by dmut** [comment]
Thanks Gareth for the PR submission, it was easy to review due to your detailed PR description and screenshots.

File: tasks/fullStack/app/helpers/reviews_helper.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    ShopRatingsCalculator.new(shop: shop).recent_monthly_changes(2)
+  end
+
+  def rating(review)
+    '&#9733;'.html_safe*review.rating.to_i
+  end
+
+  def recent_reviews_for(product)
+    product
+      .reviews
```
[the remark this answers is not in the document]
Comment (GarethPW) [reply tasks/fullStack/app/helpers/reviews_helper.rb:17]: Agree this one isn't ideal, although the per-product review limit complicates things.

I think we could use `row_number` with partitioning to bring this down to a single extra query:

```rb
columns = Arel.sql(<<~SQL)
  reviews.*, row_number() over (partition by product_id order by created_at desc)
SQL

numbered_reviews = Review
  .select(columns)
  .where(product_id: reviewed_products.ids)

Review
  .from(numbered_reviews, Review.table_name)
  .where(row_number: ..10)
  .group_by(&:product_id)
```

This is just a proof of concept and probably belongs in a query object. It's also debatable whether passing `Review.table_name` as the subquery name is a good idea here (although it *is* the only reference in the from clause).

If I'm missing an obvious, better solution please let me know!

File: tasks/fullStack/app/helpers/reviews_helper.rb (hunk 1 above)
Comment (GarethPW) [reply tasks/fullStack/app/helpers/reviews_helper.rb:28]: Oops – good spot

**Pull request closed without merging**